### PR TITLE
Simplify Metrics

### PR DESCRIPTION
**Pull request by fortuna**
This is a major simplification of the metrics logic. (~200 net lines gone!)

- stats are written by `LibevSSServer` directly to a `UsageMetricsRecorder`, which, in turn, writes to ManagerMetrics and the shared metrics.
- IP -> country is now done at `LibevSSServer`
- SharedMetrics was decomposed, so I can pass the right facet to the right code. `OutlineSharedMetricsReporter` will be able to support Prometheus by passing a Prometheus-based `OneHourUsageMetrics`
- No more writing on disk for shared metrics. Keeping the data in memory seems fine.
- LibevSSServer now has only one listener on the ss metrics socket.

After this change, I'll be able to add the Go SSServer, which reports metrics directly to Prometheus (no need for a `UsageMetricsRecorder`), and create a Prometheus-based `OneHourUsageMetrics` to pass to `OutlineSharedMetricsReporter`. The choice of which implementation to use will be done based on the roll out hash.

Tests still need to be fixed, but I think you can start taking a look.

### CODE REVIEW

File: src/shadowbox/server/main.ts [hunk 1]
```diff
@@ -37,8 +37,8 @@ const MAX_STATS_FILE_AGE_MS = 5000;
 interface MetricsConfigJson {
   // Serialized ManagerStats object.
   transferStats?: ManagerMetricsJson;
-  // Serialized SharedStats object.
-  hourlyMetrics?: SharedMetricsJson;
+  // DEPRECATED: Serialized SharedStats object.
```
Comment (trevj) [inline src/shadowbox/server/main.ts]: Is there any reason to reference this field in code? If not, let's remove this and mention it in a comment.

File: src/shadowbox/server/libev_shadowsocks_server.ts [hunk 2]
```diff
@@ -18,23 +18,61 @@ import * as dns from 'dns';
...
+      private verbose: boolean) {
+    metricsSocket.on('message', (buf: Buffer) => {
+      let metricsMessage;
+      try {
+        metricsMessage = parseMetricsMessage(buf);
+      } catch (err) {
+        logging.error('error parsing metrics: ' + buf + ', ' + err);
+        return;
+      }
+      let dataDelta = metricsMessage.totalInboundBytes -
```
Comment (trevj) [inline src/shadowbox/server/libev_shadowsocks_server.ts]:
```ts
const dataDelta = Math.max(0, ...);
```
?

File: src/shadowbox/server/libev_shadowsocks_server.ts [hunk 3]
```diff
@@ -18,23 +18,61 @@ import * as dns from 'dns';
...
+  constructor(
+      private publicAddress: string, private metricsSocket: dgram.Socket,
+      ipLocation: IpLocationService, usageRecorder: UsageMetricsRecorder,
+      private verbose: boolean) {
+    metricsSocket.on('message', (buf: Buffer) => {
+      let metricsMessage;
+      try {
+        metricsMessage = parseMetricsMessage(buf);
+      } catch (err) {
+        logging.error('error parsing metrics: ' + buf + ', ' + err);
```
Comment (trevj) [inline src/shadowbox/server/libev_shadowsocks_server.ts]: I think template strings read prettier:
```ts
 logging.error(`error parsing metrics: ${buf} (${err}`);
```

File: src/shadowbox/server/libev_shadowsocks_server.ts [hunk 4]
```diff
@@ -18,23 +18,61 @@ import * as dns from 'dns';
...
+      if (dataDelta === 0) {
+        return;
+      }
+      if (dataDelta < 0) {
+        // Counter has reset for some reason. Assume previous value was zero.
+        dataDelta = metricsMessage.totalInboundBytes;
+      }
+      this.portInboundBytes[metricsMessage.portNumber] = metricsMessage.totalInboundBytes;
+      getConnectedClientIPAddresses(metricsMessage.portNumber)
+          .then((ipAddresses: string[]) => {
```
Comment (trevj) [inline src/shadowbox/server/libev_shadowsocks_server.ts]: You shouldn't have to specify the type of `ipAddresses` here? (and below)

File: src/shadowbox/server/shared_metrics.ts [hunk 5]
```diff
@@ -12,192 +12,150 @@
...
-    this.startDatetime = new Date();
-    this.toJson(this.config.data());
-    this.config.write();
-  }
+// Keeps track of the connection metrics per user, since the startDatetime.
+// This is reported to the Outline team if the admin opts-in.
+export class OutlineSharedMetricsReporter implements SharedMetricsReporter {
+  // Date+time at which we started recording connection metrics, e.g.
+  // in case this object is constructed from data written to disk.
+  private startDatetime: Date;
```
Comment (trevj) [inline src/shadowbox/server/shared_metrics.ts]: I always thought this was an unfortunate name. It's really the time at which metrics were last reported, right?

File: src/shadowbox/server/libev_shadowsocks_server.ts (hunk 4 above)
Comment (trevj) [reply src/shadowbox/server/libev_shadowsocks_server.ts]: Sure, though plase note any decent editor will show the type.

**Comment by trevj** [comment]
@fortuna Changes looking good - let me know when you need me to TAL!

**Comment by fortuna** [comment]
Metrics id mapping done. I did some renaming along the way because I found them confusing.
I pushed to a server I have to check on the reported metrics.
Now I just need to finish the tests.

**Comment by fortuna** [comment]
Quick update: After a lot of pain I finally have tests for all the reporting and time-based behavior. I needed to add a `Clock` abstraction.

The only thing missing now is that `server_access_key.spec.ts` is using the real implementation of `LibevShadowsocksServer`, and that causes port reuse issues. I need to do a small refactoring to be able to inject a fake implementation.

@trevj If you'd like, you can take a look on all the new code other than server_access_key stuff

**Comment by fortuna** [comment]
All tests done.

@trevj This is now ready for review again.

**Pull request merged**